### PR TITLE
Remove double array allocation in SemanticTokens

### PR DESCRIPTION
**Pull request by ToddGrun**
This code was calling ArrayBuilder.ToArray, which would always allocate a new array as a result. Additionally, the arraybuilder itself would end up being large enough that it wouldn't end up going back into the pool. Thus, two large array were allocated on many calls into this method.

There really is no need for this method to use ArrayBuilder, as the code that ends up owning the resultant array doesn't attempt to put it back into a pool.

This shows up as about 0.5% of allocation in devenv in the customer profile I'm looking at, half of which should be removed by this change.

image

### CODE REVIEW

**Comment by ToddGrun** [comment]
There's an issue here, didn't notice that currentClassifiedSpanIndex was modified inside the loop, outside of normal loop incrementing

**Comment by ToddGrun** [comment]
> There's an issue here, didn't notice that currentClassifiedSpanIndex was modified inside the loop, outside of normal loop incrementing

Switched back to a pool, as the array size isn't known before the loop. Instead of ArrayBuilder, which doesn't pool array bigger than 128 elements, used a separate ObjectPool<List<int>>

**Comment by ToddGrun** [comment]
@roslyn-ide -- ptal

**Review by dibarbet — APPROVED** [review]

File: src/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 1]
```diff
@@ -243,23 +245,31 @@ private static void ConvertMultiLineToSingleLineSpans(SourceText text, Segmented
 
             var tokenTypeMap = SemanticTokensSchema.GetSchema(supportsVisualStudioExtensions).TokenTypeMap;
 
-            using var _ = ArrayBuilder<int>.GetInstance(5 * classifiedSpans.Count, out var data);
-            for (var currentClassifiedSpanIndex = 0; currentClassifiedSpanIndex < classifiedSpans.Count; currentClassifiedSpanIndex++)
+            var data = s_tokenListPool.Allocate();
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:248]: if you use .GetPooledObject you can use a `using`.

File: src/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 2]
```diff
@@ -243,23 +245,31 @@ private static void ConvertMultiLineToSingleLineSpans(SourceText text, Segmented
...
+                    data.Add(tokenType);
+                    data.Add(tokenModifiers);
+                }
 
-            return data.ToArray();
+                return data.ToArray();
+            }
+            finally
+            {
+                data.Clear();
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:270]: i am fine not clearing an list of ints.

Comment (ToddGrun) [reply src/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs]: Sure, went ahead and switched to GetPooledObject. Clearing will now occur upon retrieval from the pool.

**Pull request merged**